### DISCUSSION TITLE
FDR Filter

### DISCUSSION THREAD

**Discussion by herbiyuliantoro** [xomicsdatascience/zoDIAq#81] [Q&A]
Hello Dr. Meyer

Thanks for the wonderful software you developed for Direct Infusion. Is it possible to change the FDR filter to 0.5 or 0.1? Sometimes, people need to know protein detection with low confidence. thank you

**Comment by jessegmeyerlab** [comment]
Thanks for your interest in our work. Although I would not recommend using a higher threshold it could easily be changed. @CCranney could point you towards what line to change in your own fork. 

Caleb do you think we should add this as an option?

**Comment by CCranney** [comment]
Hi @herbiyuliantoro! Thank you for your interest. This has not been directly implemented as an end-user adjustable value, though as @jessegmeyerlab said it is quite easy to change in the code itself. I made [a branch main-fdrCutoff-0p1](https://github.com/xomicsdatascience/zoDIAq/tree/main-fdrCutoff-0p1) that changed the FDR from 0.01 to 0.1 and verified that it works on a quick manual test. In summary, you change all three instances of `fdrCutoff=0.01` to `fdrCutoff=0.1` in zodiaq -> scoring -> [fdrCalculationFunctions.py](https://github.com/xomicsdatascience/zoDIAq/blob/main-fdrCutoff-0p1/src/zodiaq/scoring/fdrCalculationFunctions.py). 

@jessegmeyerlab - At present, I do worry about adding too many low-level variables that most end users will not need or want, so do not think this should be added as an end-user option. I could see it being more confusing than helpful. I'd be interested in revisiting the issue if more use cases become apparent, though.

**Reply by herbiyuliantoro** [reply]
Thank you so much for your valuable insight. Will do it!

**Reply by herbiyuliantoro** [reply]
The reason I ask like this is because the other software has this option, like Proteome Discoverer, Maxquant, and DIA-NN. However, your software is incredible. thanks!

**Reply by CCranney** [reply]
Thank you so much! We'll take a look at those internally to see if that would be a good development direction to take. In the meantime, feel free to reach out if you have any other questions or ideas!

**Reply by herbiyuliantoro** [reply]
Hello,

Sorry just let you know that

 I have an issue with processing filtering and scoring, I copied the error code below, and then I tried to fix it, and after that, it went well. I copied the modified code below of the error code.
 
 traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\herbi\AppData\Local\Programs\Python\Python312\Scripts\zodiaq.exe\__main__.py", line 7, in <module>

  File "C:\Users\herbi\CsoDIAq\src\zodiaq\zodiaq.py", line 34, in main
    run_scoring(args)
  File "C:\Users\herbi\CsoDIAq\src\zodiaq\zodiaq.py", line 101, in run_scoring
    proteinDf = create_protein_fdr_output_from_peptide_fdr_output(peptideDf)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\herbi\CsoDIAq\src\zodiaq\scoring\fdrCalculationFunctions.py", line 52, in create_protein_fdr_output_from_peptide_fdr_output
    proteinDf["uniquePeptide"] = determine_if_peptides_are_unique_to_leading_protein(
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\herbi\CsoDIAq\src\zodiaq\scoring\fdrCalculationFunctions.py", line 191, in determine_if_peptides_are_unique_to_leading_protein
    uniquePeptides[uniqueValuesDf.index] = 1
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
IndexError: index 84 is out of bounds for axis 0 with size 82



=================================================================


def determine_if_peptides_are_unique_to_leading_protein(proteinDf):
    """
    Determines if the peptide in the peptide column is unique to the protein
        in the leadingProtein column.

    Parameters
    ----------
    proteinDf : pandas DataFrame
        The library-query match output ordered around leading proteins.

    Returns
    -------
    uniquePeptides : list
        A list containing binary values. The length of the list corresponds to the
            dataframe provided as input. 0 indicates the peptide is NOT unique, 1 that
            it is unique.
    """
    # Create an array of zeros with the same length as proteinDf
    uniquePeptides = np.zeros(len(proteinDf), dtype=int)
    # Group by leading protein and find groups with exactly one entry
    uniqueValuesDf = proteinDf.groupby("leadingProtein").filter(lambda group: len(group) == 1)
    # Set 1 for unique peptide indices, ensuring no out-of-bound index is used
    for idx in uniqueValuesDf.index:
        if idx < len(uniquePeptides):
            uniquePeptides[idx] = 1
    return uniquePeptides.tolist()
